src/test/regress/expected/local_table_join.out

```diff
@@ -0,0 +1,945 @@
+---------------------------------------------------------------------
+
+(1 row)
+
+CREATE TABLE distributed_table (key int, value text, value_2 jsonb);
+SELECT create_distributed_table('distributed_table', 'key');
+ create_distributed_table
+---------------------------------------------------------------------
+
+(1 row)
+
+CREATE TABLE distributed_table_pkey (key int primary key, value text, value_2 jsonb);
+SELECT create_distributed_table('distributed_table_pkey', 'key');
+ create_distributed_table
+---------------------------------------------------------------------
+
+(1 row)
+
+CREATE TABLE distributed_table_windex (key int primary key, value text, value_2 jsonb);
+SELECT create_distributed_table('distributed_table_windex', 'key');
+ create_distributed_table
+---------------------------------------------------------------------
+
+(1 row)
+
+CREATE UNIQUE INDEX key_index ON distributed_table_windex (key);
+CREATE TABLE distributed_partitioned_table(key int, value text) PARTITION BY RANGE (key);
+CREATE TABLE distributed_partitioned_table_1 PARTITION OF distributed_partitioned_table FOR VALUES FROM (0) TO (10);
+CREATE TABLE distributed_partitioned_table_2 PARTITION OF distributed_partitioned_table FOR VALUES FROM (10) TO (20);
+SELECT create_distributed_table('distributed_partitioned_table', 'key');
+ create_distributed_table
+---------------------------------------------------------------------
+
+(1 row)
+
+CREATE TABLE local_partitioned_table(key int, value text) PARTITION BY RANGE (key);
+CREATE TABLE local_partitioned_table_1 PARTITION OF local_partitioned_table FOR VALUES FROM (0) TO (10);
+CREATE TABLE local_partitioned_table_2 PARTITION OF local_partitioned_table FOR VALUES FROM (10) TO (20);
+CREATE TABLE distributed_table_composite (key int, value text, value_2 jsonb, primary key (key, value));
+SELECT create_distributed_table('distributed_table_composite', 'key');
+ create_distributed_table
+---------------------------------------------------------------------
+
+(1 row)
+
+SET client_min_messages TO DEBUG1;
+-- the user doesn't allow local / distributed table joinn
+SET citus.local_table_join_policy TO 'never';
+SELECT count(*) FROM postgres_table JOIN distributed_table USING(key);
+ERROR:  direct joins between distributed and local tables are not supported
+HINT:  Use CTE's or subqueries to select from local tables and use them in joins
+SELECT count(*) FROM postgres_table JOIN reference_table USING(key);
+ERROR:  direct joins between distributed and local tables are not supported
+HINT:  Use CTE's or subqueries to select from local tables and use them in joins
+-- the user prefers local table recursively planned
+SET citus.local_table_join_policy TO 'prefer-local';
+SELECT count(*) FROM postgres_table JOIN distributed_table USING(key);
+DEBUG:  Wrapping local relation "postgres_table" to a subquery: SELECT key, NULL::text AS value, NULL::jsonb AS value_2 FROM local_table_join.postgres_table WHERE true OFFSET 0
```

Comment:
as we talked in person, can we avoid `NULL`s from the target lists? It makes it hard to debug things. Also, not sure about the perf implication

Reply:
I will do some benchmark for performance here. I guess the main difficulty with getting rid of NULL's is to update all the existing range table entries, since they try to access these column (with join fields, so basically we will need to update all the join related fields in range table entries etc)

Reply:
I have created a small utility script for this purpose, feel free to play with it as well.
https://gist.github.com/SaitTalhaNisanci/103b12f59b664d4d96d879c63e86c66f


Some initial results:  (`1 selected 2 NULL`: SELECT a, NULL, NULL FROM local)
10 total cols: 1 selected, 9 NULL, 10000000 rows 0.6916604042053223 seconds
10 total cols: 1 selected, 8 NULL, 10000000 rows 0.6812009811401367 seconds
10 total cols: 1 selected, 7 NULL, 10000000 rows 0.6044158935546875 seconds
10 total cols: 1 selected, 6 NULL, 10000000 rows 0.8335869312286377 seconds
10 total cols: 1 selected, 5 NULL, 10000000 rows 0.6090021133422852 seconds
10 total cols: 1 selected, 4 NULL, 10000000 rows 0.572737455368042 seconds
10 total cols: 1 selected, 3 NULL, 10000000 rows 0.584613561630249 seconds
10 total cols: 1 selected, 2 NULL, 10000000 rows 0.7285683155059814 seconds
10 total cols: 1 selected, 1 NULL, 10000000 rows 0.6625514030456543 seconds
10 total cols: 2 selected, 8 NULL, 10000000 rows 0.6221935749053955 seconds
10 total cols: 2 selected, 7 NULL, 10000000 rows 0.8217811584472656 seconds
10 total cols: 2 selected, 6 NULL, 10000000 rows 0.900759220123291 seconds
10 total cols: 2 selected, 5 NULL, 10000000 rows 0.6378457546234131 seconds
10 total cols: 2 selected, 4 NULL, 10000000 rows 0.7322068214416504 seconds
10 total cols: 2 selected, 3 NULL, 10000000 rows 0.5630929470062256 seconds
10 total cols: 2 selected, 2 NULL, 10000000 rows 0.5395193099975586 seconds
10 total cols: 2 selected, 1 NULL, 10000000 rows 0.6461427211761475 seconds
10 total cols: 3 selected, 7 NULL, 10000000 rows 0.6598279476165771 seconds
10 total cols: 3 selected, 6 NULL, 10000000 rows 0.6342368125915527 seconds
10 total cols: 3 selected, 5 NULL, 10000000 rows 0.6533758640289307 seconds
10 total cols: 3 selected, 4 NULL, 10000000 rows 0.5480964183807373 seconds
10 total cols: 3 selected, 3 NULL, 10000000 rows 0.6522617340087891 seconds
10 total cols: 3 selected, 2 NULL, 10000000 rows 0.6625094413757324 seconds
10 total cols: 3 selected, 1 NULL, 10000000 rows 0.5635101795196533 seconds
10 total cols: 4 selected, 6 NULL, 10000000 rows 0.5155024528503418 seconds
10 total cols: 4 selected, 5 NULL, 10000000 rows 0.5159933567047119 seconds
10 total cols: 4 selected, 4 NULL, 10000000 rows 0.5202929973602295 seconds
10 total cols: 4 selected, 3 NULL, 10000000 rows 0.5519185066223145 seconds
10 total cols: 4 selected, 2 NULL, 10000000 rows 0.6474406719207764 seconds
10 total cols: 4 selected, 1 NULL, 10000000 rows 0.6520495414733887 seconds
10 total cols: 5 selected, 5 NULL, 10000000 rows 0.5719130039215088 seconds
10 total cols: 5 selected, 4 NULL, 10000000 rows 0.5545668601989746 seconds
10 total cols: 5 selected, 3 NULL, 10000000 rows 0.639406681060791 seconds
10 total cols: 5 selected, 2 NULL, 10000000 rows 0.6129019260406494 seconds
10 total cols: 5 selected, 1 NULL, 10000000 rows 0.5966455936431885 seconds
10 total cols: 6 selected, 4 NULL, 10000000 rows 0.5749166011810303 seconds
10 total cols: 6 selected, 3 NULL, 10000000 rows 0.6580791473388672 seconds
10 total cols: 6 selected, 2 NULL, 10000000 rows 0.6976237297058105 seconds
10 total cols: 6 selected, 1 NULL, 10000000 rows 0.5874862670898438 seconds
10 total cols: 7 selected, 3 NULL, 10000000 rows 0.6260466575622559 seconds
10 total cols: 7 selected, 2 NULL, 10000000 rows 0.6250591278076172 seconds
10 total cols: 7 selected, 1 NULL, 10000000 rows 0.587761640548706 seconds
10 total cols: 8 selected, 2 NULL, 10000000 rows 0.9103586673736572 seconds
10 total cols: 8 selected, 1 NULL, 10000000 rows 0.8547675609588623 seconds
10 total cols: 9 selected, 1 NULL, 10000000 rows 0.6762118339538574 seconds
10 total cols: 1 selected, 9 NULL, 100 rows 0.015694141387939453 seconds
10 total cols: 1 selected, 8 NULL, 100 rows 0.0169675350189209 seconds
10 total cols: 1 selected, 7 NULL, 100 rows 0.014212608337402344 seconds
10 total cols: 1 selected, 6 NULL, 100 rows 0.014072418212890625 seconds
10 total cols: 1 selected, 5 NULL, 100 rows 0.015524625778198242 seconds
10 total cols: 1 selected, 4 NULL, 100 rows 0.013295412063598633 seconds
10 total cols: 1 selected, 3 NULL, 100 rows 0.013313770294189453 seconds
10 total cols: 1 selected, 2 NULL, 100 rows 0.013698577880859375 seconds
10 total cols: 1 selected, 1 NULL, 100 rows 0.011465072631835938 seconds
10 total cols: 2 selected, 8 NULL, 100 rows 0.011356592178344727 seconds
10 total cols: 2 selected, 7 NULL, 100 rows 0.009987831115722656 seconds
10 total cols: 2 selected, 6 NULL, 100 rows 0.010006904602050781 seconds
10 total cols: 2 selected, 5 NULL, 100 rows 0.01069188117980957 seconds
10 total cols: 2 selected, 4 NULL, 100 rows 0.010583877563476562 seconds
10 total cols: 2 selected, 3 NULL, 100 rows 0.01146388053894043 seconds
10 total cols: 2 selected, 2 NULL, 100 rows 0.009010791778564453 seconds
10 total cols: 2 selected, 1 NULL, 100 rows 0.00992584228515625 seconds
10 total cols: 3 selected, 7 NULL, 100 rows 0.009830474853515625 seconds
10 total cols: 3 selected, 6 NULL, 100 rows 0.008640050888061523 seconds
10 total cols: 3 selected, 5 NULL, 100 rows 0.00934290885925293 seconds
10 total cols: 3 selected, 4 NULL, 100 rows 0.00909113883972168 seconds
10 total cols: 3 selected, 3 NULL, 100 rows 0.010122060775756836 seconds
10 total cols: 3 selected, 2 NULL, 100 rows 0.01024007797241211 seconds
10 total cols: 3 selected, 1 NULL, 100 rows 0.010496377944946289 seconds
10 total cols: 4 selected, 6 NULL, 100 rows 0.010824441909790039 seconds
10 total cols: 4 selected, 5 NULL, 100 rows 0.012118101119995117 seconds
10 total cols: 4 selected, 4 NULL, 100 rows 0.010517358779907227 seconds
10 total cols: 4 selected, 3 NULL, 100 rows 0.011622905731201172 seconds
10 total cols: 4 selected, 2 NULL, 100 rows 0.010773181915283203 seconds
10 total cols: 4 selected, 1 NULL, 100 rows 0.010844945907592773 seconds
10 total cols: 5 selected, 5 NULL, 100 rows 0.009765863418579102 seconds
10 total cols: 5 selected, 4 NULL, 100 rows 0.009224891662597656 seconds
10 total cols: 5 selected, 3 NULL, 100 rows 0.011211156845092773 seconds
10 total cols: 5 selected, 2 NULL, 100 rows 0.011994123458862305 seconds
10 total cols: 5 selected, 1 NULL, 100 rows 0.010563135147094727 seconds
10 total cols: 6 selected, 4 NULL, 100 rows 0.011039018630981445 seconds
10 total cols: 6 selected, 3 NULL, 100 rows 0.01609659194946289 seconds
10 total cols: 6 selected, 2 NULL, 100 rows 0.010680198669433594 seconds
10 total cols: 6 selected, 1 NULL, 100 rows 0.009272336959838867 seconds
10 total cols: 7 selected, 3 NULL, 100 rows 0.0101470947265625 seconds
10 total cols: 7 selected, 2 NULL, 100 rows 0.011170387268066406 seconds
10 total cols: 7 selected, 1 NULL, 100 rows 0.017583131790161133 seconds
10 total cols: 8 selected, 2 NULL, 100 rows 0.012769937515258789 seconds
10 total cols: 8 selected, 1 NULL, 100 rows 0.011213064193725586 seconds
10 total cols: 9 selected, 1 NULL, 100 rows 0.013700723648071289 seconds
5 total cols: 1 selected, 4 NULL, 1000000 rows 0.06916499137878418 seconds
5 total cols: 1 selected, 3 NULL, 1000000 rows 0.06649374961853027 seconds
5 total cols: 1 selected, 2 NULL, 1000000 rows 0.06927776336669922 seconds
5 total cols: 1 selected, 1 NULL, 1000000 rows 0.0681772232055664 seconds
5 total cols: 2 selected, 3 NULL, 1000000 rows 0.08529543876647949 seconds
5 total cols: 2 selected, 2 NULL, 1000000 rows 0.09929919242858887 seconds
5 total cols: 2 selected, 1 NULL, 1000000 rows 0.10389423370361328 seconds
5 total cols: 3 selected, 2 NULL, 1000000 rows 0.10135531425476074 seconds
5 total cols: 3 selected, 1 NULL, 1000000 rows 0.08748745918273926 seconds
5 total cols: 4 selected, 1 NULL, 1000000 rows 0.06591272354125977 seconds

Reply:
I guess a single run doesn't give  accurate/consistent enough results. Maybe run these two with pgbench for a longer duration:
1) 
10 total cols: 1 selected, 9 NULL, 10000000 rows
 vs
10 total cols: 1 selected, 0 NULL, 10000000 rows 

2)
5 total cols: 1 selected, 4 NULL, 100 rows
 vs
5 total cols: 1 selected, 0 NULL, 100 rows 

With  something like:

pgbench -f query.sql  -c 1 -j 1 -T 300

Reply:
We removed the NULL's in the intermediate results by doing the following trick:

SELECT t.a, NULL, NULL FROM (SELECT a FROM table)t;